File: Backend/src/main/java/com/tripnest/group/service/GroupService.java
package com.tripnest.group.service;

import com.tripnest.group.dto.CreateGroupRequest;
import com.tripnest.group.dto.InviteMemberRequest;
import com.tripnest.group.dto.TravelGroupDto;

import java.util.List;

public interface GroupService {
    TravelGroupDto createGroup(CreateGroupRequest request, String username);
    List<TravelGroupDto> getUserGroups(String username);
    TravelGroupDto getGroupDetails(Long groupId, String username);
    TravelGroupDto updateGroup(Long groupId, CreateGroupRequest request, String username);
    void deleteGroup(Long groupId, String username);
    void inviteMember(Long groupId, InviteMemberRequest request, String username);
    com.tripnest.group.dto.GroupInvitationDto inviteByEmail(Long groupId, com.tripnest.group.dto.InviteByEmailRequest request, String username);
    List<com.tripnest.group.dto.GroupInvitationDto> getPendingInvitations(String username);
    List<com.tripnest.group.dto.GroupInvitationDto> getGroupInvitations(Long groupId, String username);
    void acceptInvitation(Long invitationId, String username);
    void rejectInvitation(Long invitationId, String username);
    void cancelInvitation(Long groupId, Long invitationId, String username);
    void updateMemberRole(Long groupId, Long memberId, String role, String username);
    void transferOwnership(Long groupId, Long newAdminMemberId, String username);
    void removeMember(Long groupId, Long memberId, String username);
    void leaveGroup(Long groupId, String username);
    com.tripnest.group.dto.GroupChatMessageResponse sendGroupMessage(Long groupId, com.tripnest.group.dto.GroupChatMessageRequest request, String username);
    List<com.tripnest.group.dto.GroupChatMessageResponse> getGroupMessages(Long groupId, String username);
}

File: frontend/src/services/groupService.js
import { apiClient } from './apiClient';

export const groupService = {
  getUserGroups: async () => {
    return await apiClient.get('/groups');
  },

  getGroupDetails: async (id) => {
    return await apiClient.get(`/groups/${id}`);
  },

  createGroup: async (groupData) => {
    return await apiClient.post('/groups', groupData);
  },

  updateGroup: async (id, groupData) => {
    return await apiClient.put(`/groups/${id}`, groupData);
  },

  deleteGroup: async (id) => {
    return await apiClient.delete(`/groups/${id}`);
  },

  inviteMember: async (groupId, usernameOrEmail) => {
    return await apiClient.post(`/groups/${groupId}/members`, { usernameOrEmail, role: 'MEMBER' });
  },

  inviteByEmail: async (groupId, email, tripId) => {
    return await apiClient.post(`/groups/${groupId}/invitations/email`, { email, tripId });
  },

  getPendingInvitations: async () => {
    return await apiClient.get('/groups/invitations/pending');
  },

  getGroupInvitations: async (groupId) => {
    return await apiClient.get(`/groups/${groupId}/invitations`);
  },

  acceptInvitation: async (invitationId) => {
    return await apiClient.put(`/groups/invitations/${invitationId}/accept`);
  },

  rejectInvitation: async (invitationId) => {
    return await apiClient.put(`/groups/invitations/${invitationId}/reject`);
  },

  cancelInvitation: async (groupId, invitationId) => {
    return await apiClient.delete(`/groups/${groupId}/invitations/${invitationId}`);
  },

  updateMemberRole: async (groupId, memberId, role) => {
    return await apiClient.put(`/groups/${groupId}/members/${memberId}/role?role=${role}`);
  },

  transferOwnership: async (groupId, newAdminMemberId) => {
    return await apiClient.put(`/groups/${groupId}/transfer-ownership?newAdminMemberId=${newAdminMemberId}`);
  },

  removeMember: async (groupId, memberId) => {
    return await apiClient.delete(`/groups/${groupId}/members/${memberId}`);
  },

  leaveGroup: async (groupId) => {
    return await apiClient.post(`/groups/${groupId}/leave`);
  },

  getGroupMessages: async (groupId) => {
    return await apiClient.get(`/groups/${groupId}/messages`);
  },

  sendGroupMessage: async (groupId, message) => {
    return await apiClient.post(`/groups/${groupId}/messages`, { message });
  }
};
